examples/Custom_Hooks/probability_of_improvement.py

```diff
@@ -0,0 +1,196 @@
+
+import os
+from types import MethodType
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import torch
+from botorch.test_functions.synthetic import Hartmann
+from matplotlib.collections import PolyCollection
+from mpl_toolkits.mplot3d import Axes3D
+from scipy.stats import gaussian_kde
+
+from baybe.acquisition import ProbabilityOfImprovement
+from baybe.campaign import Campaign
+from baybe.objectives.base import Objective
+from baybe.objectives.single import SingleTargetObjective
+from baybe.parameters import NumericalDiscreteParameter
+from baybe.recommenders import (
+    BotorchRecommender,
+    RandomRecommender,
+    TwoPhaseMetaRecommender,
+)
+from baybe.searchspace import SearchSpace, SearchSpaceType
+from baybe.surrogates import GaussianProcessSurrogate
+from baybe.targets import NumericalTarget
+from baybe.utils.basic import register_hooks
+from baybe.utils.botorch_wrapper import botorch_function_wrapper
+from baybe.utils.dataframe import to_tensor
+from baybe.utils.plotting import create_example_plots
+from baybe.utils.random import set_random_seed
+
+### Settings for a full simulation loop
+
+# For the simulation, we need to define some basic settings like the number of
+# iterations or the batch size:
+
+SMOKE_TEST = "SMOKE_TEST" in os.environ
+N_DOE_ITERATIONS = 3 if SMOKE_TEST else 7
+BATCH_SIZE = 2
+DIMENSION = 3
+POINTS_PER_DIM = 2 if SMOKE_TEST else 4
+
+# We also fix the random seed to create a consistent plot:
+
+set_random_seed(1282)
+
+### Setup
+
+# We initialize a container for storing the PI from each iteration:
+
+pi_per_iteration: list[np.ndarray] = []
+
+# Then, we define the hook that calculates the PI.
+# To attach the hook, we need to match its signature to that of {meth}`RecommenderProtocol.recommend <baybe.recommenders.base.RecommenderProtocol.recommend>`.
+
+
+def extract_pi(
+    self: BotorchRecommender,
+    searchspace: SearchSpace,
+    objective: Objective | None = None,
+    measurements: pd.DataFrame | None = None,
+) -> None:
+    """Calculate and store the probability of improvement."""
+    if searchspace.type is not SearchSpaceType.DISCRETE:
+        raise TypeError(
+            f"Search spaces of type '{searchspace.type}' are not supported. "
+            f"Currently, only search spaces of type '{SearchSpaceType.DISCRETE}' are accepted."
+        )
+    train_x = searchspace.transform(measurements)
+    train_y = objective.transform(measurements)
+    acqf = ProbabilityOfImprovement()
+    botorch_acqf = acqf.to_botorch(self.surrogate_model, searchspace, train_x, train_y)
+    comp_rep_tensor = to_tensor(searchspace.discrete.comp_rep).unsqueeze(1)
+    with torch.no_grad():
+        pi = botorch_acqf(comp_rep_tensor)
+    pi_per_iteration.append(pi.numpy())
+
+
+# Additionally, we define a function that plots the collected PI values:
+
+
+def create_pi_plot(
+    pi_per_iteration: list[np.ndarray],
+) -> Axes3D:
+    """Plot the probability of improvement in 3D."""
```

Comment:
Technically, this function does not plot the PI but calculates the plot/figure/axes/... object. I'd thus rephrase this as "Create the plot of the probability ....."

Reply:
True, thanks!